File: linear-regression/linear_regression.py
import numpy as np

class LinearRegression:
    def __init__(self):
        self.w = None
        self.b = None
    
    def fit(self, X, y):
        X_mean = np.mean(X)
        y_mean = np.mean(y)
        self.w = np.sum((X - X_mean) * (y - y_mean)) / np.sum((X - X_mean) ** 2)
        self.b = y_mean - self.w * X_mean
    
    def predict(self, X):
        return self.w * X + self.b

# Test
X = np.array([1, 2, 3, 4, 5])
y = np.array([2, 4, 6, 8, 10])
model = LinearRegression()
model.fit(X, y)
print(model.predict([6]))  # [12]

File: linear-regression/linear_regression.clj
(defn linear-regression [X y]
  (let [X-mean (/ (reduce + X) (count X))
        y-mean (/ (reduce + y) (count y))
        num (reduce + (map #(* (- %1 X-mean) (- %2 y-mean)) X y))
        den (reduce + (map #(* (- % X-mean) (- % X-mean)) X))
        w (/ num den)
        b (- y-mean (* w X-mean))]
    {:w w :b b}))

(defn predict [model x]
  (+ (* (:w model) x) (:b model)))

(let [model (linear-regression [1 2 3 4 5] [2 4 6 8 10])]
  (println (predict model 6))) ; 12.0
